### PR TITLE
fix(settings): fixed two IPFS gateway issues

### LINKED ISSUE #16871: [Bug]: IPFS Gateway input not validated and seeing the encoded value on Advanced Settings
### Describe the bug

**Problem**: the IPFS Gateway input seems not validated as I can introduce spaces and any char as well. Once I've saved the value on the Onboarding screen, I see that the Advanced Settings displayed the encoded version for some chars like spaces as %20.

Notice the lack of validation seems it's an old problem, already happening on the Advanced Settings.

https://user-images.githubusercontent.com/54408225/206535323-10d85a4a-e632-4ede-8fa9-4fc0eced3e31.mp4

### Steps to reproduce

1. Load MM for the first time
2. Click Create Wallet and proceed with the onboarding
3. Click Advanced Settings
4. Add an invalid IPFS gateway "     mdsljkd   sdfs"
5. Proceed
6. Land at Home
7. Go to Advanced Settings -- see how input is encoded

### Error messages or log output

_No response_

### Version

10.24.0

### Build type

None

### Browser

Chrome

### Operating system

Linux

### Hardware wallet

_No response_

### Additional context

_No response_

### LINKED ISSUE #18140: [Bug]: When editing the IPFS Gateway URL - land on MetaMask encountered an error page
### Describe the bug

**Problem**: when I go to Privacy Settings, if I try to edit the IPFS Gateway option I land into the MM encountered an error page saying `Invalid URL`.

https://user-images.githubusercontent.com/54408225/224991238-b8cb55bc-e03c-456f-97e2-1dd418985601.mp4

### Steps to reproduce

1. Go to Privacy Settings
2. Edit the IPFS Gateway to either blank or `http://` or `https://`
3. Land on MM encountered an error page

### Error messages or log output

_No response_

### Version

10.25.0

### Build type

None

### Browser

Chrome

### Operating system

Linux

### Hardware wallet

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by HowardBraham**
## Explanation

- adds back in two bugfixes that were originally in #19283
  - fixes #16871
  - fixes #18140

- achieves 100% code coverage for /ui/pages/settings/security-tab
- removes the npm package `valid-url`, which has not been updated in 10 years

### CODE REVIEW

**Review by NidhiKJha — COMMENTED** [review]
Valid-url removal is really nice. Though I can still enter IPFS with spaces. We need to validate that?



https://github.com/MetaMask/metamask-extension/assets/39872794/17447f5f-7d6c-4969-bf4d-3c7252983c21

**Comment by HowardBraham** [comment]
> Valid-url removal is really nice. Though I can still enter IPFS with spaces. We need to validate that?

This is a tricky situation, because a URL actually **can** have a `%20` space in it.  Is it done a lot?  No.

How do we handle this in other places?

**Comment by Gudahtt** [comment]
Spaces are allowed in the path, but not the origin. We need to retain some amount of validation here to protect users from accidentally entering invalid URLs.

**Comment by HowardBraham** [comment]
@Gudahtt it's actually already validating through URL

Try `new URL('https://a b')`, it doesn't produce an error

I'm not actually sure if `https://a%20b.com/` is a valid URL, I can't find any clear documentation

**Comment by HowardBraham** [comment]
@Gudahtt @NidhiKJha update... you cannot have a domain name with an encoded space in it because `%`  is not a valid character.  But... URL behaves differently in Node and browser!

- Node `new URL('https://a b.com')` gives an error
- Browser `new URL('https://a b.com')` is fine

Options:
1. Validate with another package, maybe `is-url-http` or `url`
2. Validate with a regex

Any preferences?

**Comment by HowardBraham** [comment]
@Gudahtt @NidhiKJha I fixed the URL validation this way:

```
const url = new URL(urlString);

if (url.hostname !== decodeURIComponent(url.hostname)) {
  return null; // will happen if there's a %, a space, or other invalid character in the hostname
}
```

It also now incorporates the changes from #20172

Hopefully should be good to approve and merge now

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js [hunk 1]
```diff
@@ -83,27 +83,14 @@ async function addEthereumChainHandler(
     );
   }
 
-  const isLocalhost = (strUrl) => {
```
[the remark this answers is not in the document]
Comment (HowardBraham) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:86]: Yeah I see what you mean... it's better, but not good enough to put in util.  I moved the new version of it back to `add-ethereum-chain.js`.

File: app/scripts/lib/util.ts [hunk 2]
```diff
@@ -235,10 +234,54 @@ export function previousValueComparator<A>(
 }
 
 export function addUrlProtocolPrefix(urlString: string) {
-  if (!urlString.match(/(^http:\/\/)|(^https:\/\/)/u)) {
-    return `https://${urlString}`;
+  let trimmed = urlString.trim();
+
+  if (trimmed.length > 0 && !trimmed.match(/(^http:\/\/)|(^https:\/\/)/u)) {
```
[the remark this answers is not in the document]
Comment (HowardBraham) [reply app/scripts/lib/util.ts]: Good suggestion, but didn't quite work the way you wrote it.  The problem is that `new URL('example.com')` throws an error.  I rewrote this line to
`if (trimmed.length && !url.parse(trimmed).protocol) {`

**Review by NidhiKJha — DISMISSED** [review]

**Review by NidhiKJha — APPROVED** [review]

File: app/scripts/lib/util.ts [hunk 3]
```diff
@@ -235,10 +235,43 @@ export function previousValueComparator<A>(
...
-  if (!urlString.match(/(^http:\/\/)|(^https:\/\/)/u)) {
-    return `https://${urlString}`;
+  let trimmed = urlString.trim();
+
+  if (trimmed.length && !urlLib.parse(trimmed).protocol) {
+    trimmed = `https://${trimmed}`;
   }
-  return urlString;
+
+  if (getValidUrl(trimmed) !== null) {
```
Comment (montelaidev) [inline app/scripts/lib/util.ts:244]: It looks like we need to add additional tests for edge cases like this 

```
..
.asdf.
.asdf
adsf.
```
  If we prepend the http:// before getValidUrl these cases would be valid.

Comment (montelaidev) [reply app/scripts/lib/util.ts:244]: As discussed with @HowardBraham, we can handle these edge cases in another pr.

**Review by montelaidev — APPROVED** [review]

**Pull request merged**